File: ParProg-Testat-2/Aufgabe2/Java/aufgabe2/BankAccount.java
package aufgabe2;

import java.util.concurrent.atomic.AtomicInteger;

public class BankAccount {
    private AtomicInteger balance = new AtomicInteger(0);

    public void deposit(int amount) {
        balance.addAndGet(amount);
    }

    public boolean withdraw(int amount) {
        int oldValue, newValue;
        do {
            oldValue = balance.get();
            if (oldValue < amount) {
                return false;
            }
            newValue = oldValue - amount;
        } while (!balance.compareAndSet(oldValue, newValue));
        return true;
    }

    public int getBalance() {
        return balance.get();
    }
}

File: ParProg-Testat-2/Aufgabe2/DotNet/BankAccount/BankAccount.cs
using System.Threading;

namespace BankAccount
{
    public class BankAccount
    {
        private int _balance = 0;

        public int Balance => _balance;

        public void Deposit(int amount)
        {
            Interlocked.Add(ref _balance, amount);
        }

        public bool Withdraw(int amount)
        {
            int initialValue, computedValue;
            do
            {
                initialValue = _balance;
                if (initialValue < amount)
                { 
                    return false;
                }

                computedValue = initialValue - amount;
            } while (initialValue != Interlocked.CompareExchange(ref _balance, computedValue, initialValue));

            return true;
        }
    }
}